sdk/core/azure_core_amqp/src/connection.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corp. All Rights Reserved.
+// cspell: words amqp sasl
+
+use super::value::{AmqpOrderedMap, AmqpSymbol, AmqpValue};
+use azure_core::error::Result;
+use std::fmt::Debug;
+use time::Duration;
+use url::Url;
+
+pub struct AmqpConnectionOptions {
+    pub(crate) max_frame_size: Option<u32>,
+    pub(crate) channel_max: Option<u16>,
+    pub(crate) idle_timeout: Option<Duration>,
+    pub(crate) outgoing_locales: Option<Vec<String>>,
+    pub(crate) incoming_locales: Option<Vec<String>>,
+    pub(crate) offered_capabilities: Option<Vec<AmqpSymbol>>,
+    pub(crate) desired_capabilities: Option<Vec<AmqpSymbol>>,
+    pub(crate) properties: Option<AmqpOrderedMap<AmqpSymbol, AmqpValue>>,
+    pub(crate) buffer_size: Option<usize>,
+}
+
+impl AmqpConnectionOptions {
+    pub fn builder() -> builders::AmqpConnectionOptionsBuilder {
+        builders::AmqpConnectionOptionsBuilder::new()
+    }
+}
+
+pub trait AmqpConnectionTrait {
+    #[allow(unused_variables)]
+    fn open(
+        &self,
+        name: impl Into<String>,
+        url: Url,
+        options: Option<AmqpConnectionOptions>,
+    ) -> impl std::future::Future<Output = Result<()>> {
+        async {
+            unimplemented!();
+        }
+    }
+    fn close(&self) -> impl std::future::Future<Output = Result<()>> {
+        async {
+            unimplemented!();
+        }
+    }
+}
+
+#[cfg(all(feature = "enable-fe2o3-amqp", not(target_arch = "wasm32")))]
+type ConnectionImplementation = super::fe2o3::connection::Fe2o3AmqpConnection;
+
+#[cfg(any(not(feature = "enable-fe2o3-amqp"), target_arch = "wasm32"))]
+type ConnectionImplementation = super::noop::NoopAmqpConnection;
+
+#[derive(Debug, Default)]
+pub(crate) struct AmqpConnectionImpl<T>(pub(crate) T);
+
+impl<T> AmqpConnectionImpl<T>
+where
+    T: AmqpConnectionTrait,
+{
+    pub(crate) fn new(connection: T) -> Self {
+        Self(connection)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct AmqpConnection(pub(crate) AmqpConnectionImpl<ConnectionImplementation>);
+
+impl AmqpConnectionTrait for AmqpConnection {
```

Comment:
I don't understand the point of this trait. Could you explain? There's a lot of traits here, some of which I hope aren't necessary. .NET likes to use lots of interfaces. Rust does not (well, traits, but similar).

Reply:
There's basically 1 trait per construct, and 3 implementations of the trait. One implementation is a thin forwarder to the actual implementation (that's the one you see in this file). The two implementations are in noop.rs (that's the "NoOp implementation which fails everything"), and the one in the fe2o3 module (which is the "real" iron oxide client).

It's a trait because I need all 3 implementations to have the same API surface. And I don't want to expose any of the feature complexity outside the AMQP package - that's why I have two newtypes - the concrete one in AmqpConnection and the generic one in AmqpConnectionImpl which is specialized in AmqpConnection.

Reply:
Ok, now a trait and 3 implementations of the trait, removing both of the newtypes.